README.md
# Segment-Nielsen-DCR

[![Circle CI](https://circleci.com/gh/segment-integrations/analytics-ios-integration-nielsen-dcr.svg?style=svg&circle-token=9df3a0f8385bd1f43655f79cf649035cfe538035)](https://circleci.com/gh/segment-integrations/analytics-ios-integration-nielsen-dcr)
[![Version](https://img.shields.io/cocoapods/v/Segment-Nielsen-DCR.svg?style=flat)](http://cocoapods.org/pods/Segment-Nielsen-DCR)
[![License](https://img.shields.io/cocoapods/l/Segment-Nielsen-DCR.svg?style=flat)](http://cocoapods.org/pods/Segment-Nielsen-DCR)
[![Platform](https://img.shields.io/cocoapods/p/Segment-Nielsen-DCR.svg?style=flat)](http://cocoapods.org/pods/Segment-Nielsen-DCR)

## Installation

The Nielsen App SDK as of version 6.0.0.0 is compatible with Apple iOS version 8.0 and above.

Segment-Nielsen-DCR SDK is not available through [CocoaPods](http://cocoapods.org) trunk due to Nielsen's SDK being in a private spec repo. To install the Segment-Nielsen-DCR pod, add the following line to your Podfile:

```ruby
pod "Segment-Nielsen-DCR", :git => 'https://github.com/segment-integrations/analytics-ios-integration-nielsen-dcr.git', :tag => '1.5.1'
```

The integration relies on the the Nielsen framework, which can either be installed via Cocoapods or by manually adding the framework. You will need to have a Nielsen representative to get started.

## Xcode 12

The Nielsen Cocoapod (8.0.0.0) specifies iOS 8.0 as it's minimum target.  This has the negative effect of generating ones Cocoapod workspace with invalid values.  To fix this, the following script can be added to the end of the `Podfile`:

```
post_install do |installer|
  installer.pods_project.targets.each do |target|
    target.build_configurations.each do |config|
      config.build_settings.delete 'IPHONEOS_DEPLOYMENT_TARGET'
    end
  end
end
```

This will remove the deployment target settings and fall back to what has been specified within the Xcode Project itself.

It should also be noted that the Nielsen Cocoapod does not include an `arm64` slice.  It may be necessary for  `EXCLUDED_ARCHS[sdk=iphonesimulator*]` to be set to exclude `arm64` in your project.

## Cocoapods

When using the Nielsen SDK version 6.2.0.0 and above, Nielsen recommends installation via Cocoapods, and Apple recommends using the dynamic framework.

Installation of the Dynamic Nielsen App Framework SDK via Cocoapods requires a Cocoapods version 1.6.1 or higher. Installation of the Static Nielsen App Framework SDK via Cocoapods requires a Cocoapods version 1.4.0 or higher.

1. **Set your repository credentials.** The first step is to add the credentials received from Nielsen into your `.netrc` file. Navigate to your home folder and create a file called `.netrc`
```
cd ~/
vi .netrc
```
You will need to fill out a license agreement form and have the contact information for your Nielsen representative in order to obtain the credentials [here](https://engineeringportal.nielsen.com/docs/Special:Downloads). Add the credentials in the following format:
```
machine raw.githubusercontent.com
login <Nielsen App SDK client>
password <Auth token>
```

2. **Add the source to your Podfile**

  Dynamic Framework (Note: you will need to include `use_frameworks!`)
```
source 'https://github.com/NielsenDigitalSDK/nielsenappsdk-ios-specs-dynamic.git'
```
Static Framework
```
source 'https://github.com/NielsenDigitalSDK/nielsenappsdk-ios-specs.git'
```

3. **Add the pod to your Podfile**

  `pod NielsenAppSDK`

4. **Install the pods**

  `pod install`

***A list of full instructions can be found [here](https://engineeringportal.nielsen.com/docs/Digital_Measurement_iOS_Artifactory_Guide)***

## Manual

Navigate to [Nielsen's Engineering Site](https://engineeringportal.nielsen.com/docs/Main_Page) and download the following Video framework:

![](http://g.recordit.co/IvvLm8oAY2.gif)

There will be an NDA to sign prior to accessing the download. Nielsen requires you fill out your company info and have a Nielsen representative before getting started.

Nielsen also requires the following frameworks, which must be included into Link Binary with Libraries (within app target’s Build Phases)
  - AdSupport.framework
  - SystemConfiguration.framework
  - CoreLocation.framework (Not applicable for International (Germany))
  - libsqlite3

## Usage

After you have properly installed the Nielsen App SDK, Register the factory with Segment SDK in the `application:didFinishLaunchingWithOptions`  method of your `AppDelegate`:

`#import <Segment-Nielsen-DCR/SEGNielsenDCRIntegrationFactory.h>`


```
NSString *const SEGMENT_WRITE_KEY = @" ... ";
SEGAnalyticsConfiguration *config = [SEGAnalyticsConfiguration configurationWithWriteKey:SEGMENT_WRITE_KEY];

[config use:[SEGNielsenDCRIntegrationFactory instance]];

[SEGAnalytics setupWithConfiguration:config];
```

## License

Segment-Nielsen-DCR is available under the MIT license. See the LICENSE file for more info.


CHANGELOG.md
Change Log
==========

Version 1.5.1 *(11th July, 2022)*
-------------------------------------------
* Added assetid to Nielsen properties in Screen calls, pulled from the asset_id property

Version 1.5.0 *(7th April, 2021)*
-------------------------------------------
* Updates `Video Playback Started` to call `play` and load content metadata.
* Updates `Video Playback Interrupted` to call `stop`.
* Maps `Video Playback Exited` to call `stop`.
* Removes `+1` increment to playhead when `position` = 0.
* Adds support for setting to map custom property to Nielsen's page/screen `section`, with fallback to `name` field.

Version 1.4.0 *(9th October, 2020)*
-------------------------------------------
* Updates Nielsen SDK to 8.x
* Fixes out-of-the-box compilation issues when used as a pod.

Version 1.3.3 *(7th October, 2020)*
-------------------------------------------
* Update SEGAnalytics import to support new namespacing introduced in v4.x

Version 1.3.2 *(21st June, 2020)*
-------------------------------------------
* Relaxes Segment Analytics library dependency.

Version 1.3.1 *(19th May, 2020)*
-------------------------------------------
* Updates `returnPlayheadPosition` to set the `playheadPosition` to the  `currentTime` when the new setting `sendCurrentTimeLivestream` is enabled on livestream videos.

Version 1.3.0 *(18th October, 2019)*
-------------------------------------------
* [DEST-1241]
* Supports playback events "Video Playback Seek Started", "Video Playback   Buffer Started" and "Video Playback Buffer Completed".
* sfCode setting hard-coded to "dcr" per Nielsen request.
* Timer supports livestream offset as value of "position" property.

Version 1.2.8 *(17th September, 2019)*
-------------------------------------------
* Updates `returnAdLoadType` helper method to check for a load type in `properties.load_type`.

Version 1.2.7 *(9th September, 2019)*
-------------------------------------------
* Updates `returnAdLoadType` helper method to check for a load type in `properties.loadType`. Will fallback to `adLoadType` in integrations specific options.

Version 1.2.6 *(30th August, 2019)*
-------------------------------------------
* Adds a helper functions `returnAirdate` that will convert a date passed in ISO 8601 format or in format 2019-30-08 to Nielsen yyyyMMdd HH:MM:SS `airdate` format.
* Removes hyphens for ad `type` of "post-roll" and "mid-roll"

Version 1.2.5 *(14th August, 2019)*
-------------------------------------------
* Fixes a bug in helper functions setting nil values in contentMetadata.
* Adds support for retrieving `optOutURL`.

Version 1.2.4 *(5th August, 2019)*
-------------------------------------------
* Fixes a bug with `content_length` helper function
* Updates pre-roll ad loadMetadata. Previously we were retrieving properties for content on  pre-roll ads from `properties.content` to adhere to the Segment video spec.

Version 1.2.3 *(25th July, 2019)*
-------------------------------------------
* Adds support for Nielsen's ad `assetId` to be  assigned using the setting: `adAssetIdPropertyName` and custom property mapping.
* Changes naming convention of `assetIdPropertyName` to `contentAssetIdPropertyName` for custom content assetId mapping.

Version 1.2.2 *(25th July, 2019)*
-------------------------------------------
* Fixes param values passed to `returnContentLength`.

Version 1.2.1 *(19th July, 2019)*
-------------------------------------------
* Adds functionality support setting custom length property ("contentLengthPropertyName") in Segment UI. This is mapped to Nielsen's length. If a property is not provided in the settings, Segment will default to Segment specced property, `total_length`.

Version 1.2.0 *(17th July, 2019)*
-------------------------------------------
* Adds functionality support setting custom assetid ("assetIdPropertyName") in Segment UI. This is used to override default property mapping for Segment property 'asset_id' to Nielsen 'assetid' when present.
* Adds functionality support setting custom clientid ("clientIdPropertyName") in Segment UI. This is used to override the 'cid'/'clientid' value sent to Nielsen in 'loadMetadata()' events.
* Adds functionality support setting custom subbrand ("subbrandPropertyName") in Segment UI. This is used to override the 'vcid'/'subbrand' value sent to Nielsen in 'loadMetadata()' events.

Version 1.1.2 *(16th July, 2019)*
-------------------------------------------
* Fix syntax errors of Content Metadata and AppInformation.

Version 1.1.1 *(15th July, 2019)*
-------------------------------------------
* Fix casing of mediaUrl and channelName to have  parity with Android.

Version 1.1.0 *(11th July, 2019)*
-------------------------------------------
* Add debug flag/setting
* Migrates from travis to CircleCI
* Updates build and Readme per SDK bundling
* Adds `hasAds` and `crossId2` to content metadata

Version 1.0.0-beta *(13th September, 2017)*
-------------------------------------------
*(Supports analytics-ios 3.6.+ and Nielsen-DCR 5.1.1.25+)*

Initial release.


Segment-Nielsen-DCR/Classes/SEGNielsenDCRIntegration.m
//
//  SEGNielsenDCRIntegration.m
//  Pods
//
//  Created by ladan nasserian on 6/19/17.
//
//

#import "SEGNielsenDCRIntegration.h"
#if defined(__has_include) && __has_include(<Analytics/SEGAnalytics.h>)
#import <Analytics/SEGAnalyticsUtils.h>
#else
#import <Segment/SEGAnalyticsUtils.h>
#endif
#import <NielsenAppApi/NielsenAppApi.h>

#pragma mark -
#pragma mark Property Conversion
#pragma mark -

NSString *returnFullEpisodeStatus(NSDictionary *src, NSString *key)
{
    NSNumber *value = [src valueForKey:key];
    if ([value isEqual:@YES]) {
        return @"y";
    }
    return @"n";
}


NSString *returnAdLoadType(NSDictionary *options, NSDictionary *properties)
{
    NSString *value;
    if ([options valueForKey:@"adLoadType"]){
        value = [options valueForKey:@"adLoadType"];
    } else if ([properties valueForKey:@"loadType"]){
        value = [properties valueForKey:@"loadType"];
    } else if ([properties valueForKey:@"load_type"]){
        value = [properties valueForKey:@"load_type"];
    }
    
    if ([value isEqualToString:@"dynamic"]) {
        return @"2";
    }
    return @"1";
}

NSString *returnHasAdsStatus(NSDictionary *src, NSString *key)
{
    NSNumber *value = [src valueForKey:key];
    if ([value isEqual:@YES]) {
        return @"1";
    }
    return @"0";
}

NSString *returnContentLength(NSDictionary *src, NSString *defaultKey, NSDictionary *settings)
{
    NSString *contentLengthKey = settings[@"contentLengthPropertyName"];
    NSString *contentLength;
    NSString *customContentLength = src[contentLengthKey];
    if ([contentLengthKey length] > 0 && customContentLength) {
        contentLength = [src valueForKey:contentLengthKey];
    } else if (src[@"total_length"]) {
       contentLength = [src valueForKey:@"total_length"];
    } else {
        contentLength = @"";
    }
    
    return contentLength;
}

NSString *returnCustomContentAssetId(NSDictionary *properties, NSString *defaultKey, NSDictionary *settings)
{
    NSString *customKey = settings[@"contentAssetIdPropertyName"];
    NSString *value;
    NSString *customContentAssetId = properties[customKey];
    if ([customKey length] > 0 && customContentAssetId){
        value = [properties valueForKey:customKey];
    } else if (properties[defaultKey]) {
        value = [properties valueForKey:defaultKey];
    } else {
        value = @"";
    }
    return value;
}

NSString *returnCustomAdAssetId(NSDictionary *properties, NSString *defaultKey, NSDictionary *settings)
{
    NSString *customKey = settings[@"adAssetIdPropertyName"];
    NSString *value;
    NSString *customAssetId = properties[customKey];
    if ([customKey length] > 0 && customAssetId){
        value = [properties valueForKey:customKey];
    } else if (properties[defaultKey])  {
        value = [properties valueForKey:defaultKey];
    } else {
        value = @"";
    }
    return value;
}

NSString *returnCustomSectionProperty(NSDictionary *properties, NSString *defaultKey, NSDictionary *settings)
{
    NSString *customKey = settings[@"customSectionProperty"];
    NSString *value;
    NSString *customSectionName = properties[customKey];
    if ([customKey length] > 0 && customSectionName){
        value = [properties valueForKey:customKey];
    } else if (defaultKey)  {
        value = defaultKey;
    } else {
        value = @"Unknown";
    }
    return value;
}

NSString *returnAirdate(NSDictionary *properties, NSString *defaultKey)
{
        NSString *dateSTR = [properties valueForKey:defaultKey];
        NSArray *dateArrayISO = [dateSTR componentsSeparatedByString:@"T"];
        NSArray *dateArraySimpleDate = [dateSTR componentsSeparatedByString:@"-"];
        NSDate *date;
        //Convert the string date if it was passed in ISO 8601 format ie. 2019-08-30T21:00:00Z
        if ([dateArrayISO count] > 1) {
            NSISO8601DateFormatter *ISODateFormatter = [[NSISO8601DateFormatter alloc] init];
            date = [ISODateFormatter dateFromString:dateSTR];
        }
    
        //Convert the string date if it was passed in simple date format ie. 2019-08-30
        if ([dateArraySimpleDate count] > 2 && [dateSTR length] == 10) {
            NSDateFormatter *simpleDateFormatter = [[NSDateFormatter alloc] init];
            [simpleDateFormatter setDateFormat:@"yyyy-MM-dd"];
            NSTimeZone *timeZone = [NSTimeZone timeZoneWithName:@"UTC"];
            [simpleDateFormatter setTimeZone:timeZone];
            date = [simpleDateFormatter dateFromString:dateSTR];
        }
    
        //Manipulate the date to Nielsen format
        NSDateFormatter *nielsenDateFormatter = [[NSDateFormatter alloc] init];
        NSTimeZone *timeZone = [NSTimeZone timeZoneWithName:@"UTC"];
        [nielsenDateFormatter setTimeZone:timeZone];
        [nielsenDateFormatter setDateFormat:@"yyyyMMdd HH:mm:ss"];
        NSString *nielsenDateString = [nielsenDateFormatter stringFromDate:date];
    
        if ([nielsenDateString length] > 0) {
            return nielsenDateString;
        } else if ([dateSTR length] > 0) {
            return dateSTR;
        } else {
            return @"";
        }
}

long long returnPlayheadPosition(SEGTrackPayload *payload, NSDictionary *settings)
{
    long long playheadPosition = 0;
    

    NSDictionary *properties = payload.properties;
    // if livestream, you need to send current UTC timestamp
    if ([properties[@"livestream"] isEqual:@YES]) {
        long long position = 0;
        position = [properties[@"position"] longLongValue];
        long long currentTime = [[NSDate date] timeIntervalSince1970];
        if ([settings[@"sendCurrentTimeLivestream"] isEqual:@YES]){
            //for livestream, if this setting is enabled just send the curent time
            playheadPosition = currentTime;
        } else {
            //for livestream, properties.position is a negative integer representing offset in seconds from current time
            playheadPosition = currentTime + position;
        }
    } else if (properties[@"position"]) {
        // if position is passed in we should override the state of the counter with the explicit position given from the customer
        playheadPosition = [properties[@"position"] longLongValue];
    };

    return playheadPosition;
}

// Nielsen expects all value type String.
NSDictionary *coerceToString(NSDictionary *map)
{
    NSMutableDictionary *newMap = [NSMutableDictionary dictionaryWithDictionary:map];

    for (id key in map) {
        id value = [map objectForKey:key];
        if (![value isKindOfClass:[NSString class]]) {
            [newMap setObject:[NSString stringWithFormat:@"%@", value] forKey:key];
        }
    }

    return [newMap copy];
}

#pragma mark -
#pragma mark Metadata Mapping
#pragma mark -

NSDictionary *returnMappedContentProperties(NSDictionary *properties, NSDictionary *options, NSDictionary *settings)
{
    NSDictionary *contentMetadata = @{
        @"pipmode" : options[@"pipmode"] ?: @"false",
        @"adloadtype" : returnAdLoadType(options, properties),
        @"assetid" : returnCustomContentAssetId(properties, @"asset_id", settings),
        @"type" : @"content",
        @"segB" : options[@"segB"] ?: @"",
        @"segC" : options[@"segC"] ?: @"",
        @"title" : properties[@"title"] ?: @"",
        @"program" : properties[@"program"] ?: @"",
        @"isfullepisode" : returnFullEpisodeStatus(properties, @"full_episode"),
        @"hasAds" : returnHasAdsStatus(options, @"hasAds"),
        @"airdate" : returnAirdate(properties, @"airdate"),
        @"length" : returnContentLength(properties, @"content_length", settings),
        @"crossId1" : options[@"crossId1"] ?: @"",
        @"crossId2" : options[@"crossId2"] ?: @""
    };

    NSMutableDictionary *mutableContentMetadata = [contentMetadata mutableCopy];
    if (settings[@"subbrandPropertyName"]){
        NSString *subbrandValue = properties[settings[@"subbrandPropertyName"]] ?: @"";
        [mutableContentMetadata setObject:subbrandValue forKey:@"subbrand"];
    }

    if (settings[@"clientIdPropertyName"]){
        NSString *clientIdValue = properties[settings[@"clientIdPropertyName"]] ?: @"";
        [mutableContentMetadata setObject:clientIdValue forKey:@"clientid"];
    }

    return coerceToString(mutableContentMetadata);
}

NSDictionary *returnMappedAdProperties(NSDictionary *properties, NSDictionary *options, NSDictionary *settings)
{
    NSDictionary *adMetadata = @{
        @"assetid" : returnCustomAdAssetId(properties, @"asset_id", settings),
        @"type" : properties[@"type"] ?: @"ad",
        @"title" : properties[@"title"] ?: @""

    };
    
    NSMutableDictionary *mutableAdMetadata = [adMetadata mutableCopy];

    if ([properties[@"type"] isEqualToString:@"pre-roll"]) {
        [mutableAdMetadata setObject:@"preroll" forKey:@"type"];
    }
    
    if ([properties[@"type"] isEqualToString:@"mid-roll"]) {
        [mutableAdMetadata setObject:@"midroll" forKey:@"type"];
    }
    
    if ([properties[@"type"] isEqualToString:@"post-roll"]) {
        [mutableAdMetadata setObject:@"postroll" forKey:@"type"];
    }
    
    return coerceToString(mutableAdMetadata);
}

#pragma mark - Integration


@interface SEGNielsenDCRIntegration () <NielsenAppApiDelegate>

@end


@implementation SEGNielsenDCRIntegration

- (instancetype)initWithSettings:(NSDictionary *)settings andNielsen:(NielsenAppApi *)nielsen
{
    if (self = [super init]) {
        self.settings = settings;
        self.nielsen = nielsen;

        NSString *appName = [[NSBundle mainBundle] objectForInfoDictionaryKey:@"CFBundleName"];
        NSString *appVersion = [[NSBundle mainBundle] objectForInfoDictionaryKey:@"CFBundleVersion"];

        NSMutableDictionary *appInformation = [[NSMutableDictionary alloc] initWithDictionary: @{
            @"appid" : settings[@"appId"] ?: @"",
            @"appname" : appName ?: @"",
            @"appversion" : appVersion ?: @"",
            @"sfcode" : @"dcr" // hard-coding sfcode to "dcr" per request from Nielsen's support team
        }];

        if ([settings[@"nolDevDebug"] boolValue]) {
            [appInformation addEntriesFromDictionary:@{@"nol_devDebug": @"DEBUG"}];
        }

        if (nielsen == nil) {
            self.nielsen = [[NielsenAppApi alloc] initWithAppInfo:appInformation delegate:nil];
            SEGLog(@"[NielsenAppApi alloc] initWithAppInfo: %@ delegate:nil]", appInformation);
        }
    }

    return self;
}

#pragma mark -
#pragma mark Heartbeat Timers
#pragma mark -

- (void)playHeadTimeEvent:(NSTimer *)timer
{ 
    self.startingPlayheadPosition = self.startingPlayheadPosition + 1;

    [self.nielsen playheadPosition:self.startingPlayheadPosition];
    SEGLog(@"[NielsenAppApi playheadPosition: %d]", self.startingPlayheadPosition);
}


- (void)startPlayheadTimer:(SEGTrackPayload *)payload
{
    NSDictionary *settings = self.settings;
    dispatch_async(dispatch_get_main_queue(), ^{
        if (self.playheadTimer == nil) {
            // Remove 1 from playhead position to maintain original position
            self.startingPlayheadPosition = returnPlayheadPosition(payload, settings) - 1;
            self.playheadTimer = [NSTimer scheduledTimerWithTimeInterval:1.0 target:self selector:@selector(playHeadTimeEvent:) userInfo:nil repeats:YES];
        }
    });
}

- (void)stopPlayheadTimer:(SEGTrackPayload *)payload
{
    dispatch_async(dispatch_get_main_queue(), ^{
        [self.nielsen playheadPosition:self.startingPlayheadPosition];
        SEGLog(@"[NielsenAppApi playheadPosition: %d]", self.startingPlayheadPosition);

        if (self.playheadTimer) {
            [self.playheadTimer invalidate];
            self.playheadTimer = nil;
        }
    });
}

#pragma mark - Track

- (void)track:(SEGTrackPayload *)payload
{
    NSDictionary *properties = payload.properties;
    NSDictionary *options = [payload.integrations valueForKey:@"nielsen-dcr"];
#pragma mark Playback Events

    // Nielsen requires we load content metadata and call play upon playback start
    if ([payload.event isEqualToString:@"Video Playback Started"]) {
        NSDictionary *channelInfo = @{
            // channelName is optional for DCR, if not present Nielsen asks to set default
            @"channelName" : options[@"channelName"] ?: @"defaultChannelName",
            // if mediaURL is not available, Nielsen expects an empty value
            @"mediaURL" : options[@"mediaUrl"] ?: @""
        };
        NSDictionary *contentMetadata = returnMappedContentProperties(properties, options, self.settings);
        [self.nielsen loadMetadata:contentMetadata];
        SEGLog(@"[NielsenAppApi loadMetadata:%@]", contentMetadata);
        [self startPlayheadTimer:payload];
        [self.nielsen play:channelInfo];
        SEGLog(@"[NielsenAppApi play: %@]", channelInfo);
        return;
    }

    if ([payload.event isEqualToString:@"Video Playback Resumed"] ||
        [payload.event isEqualToString:@"Video Playback Seek Completed"] ||
        [payload.event isEqualToString:@"Video Playback Buffer Completed"]) {
        NSDictionary *channelInfo = @{
            // channelName is optional for DCR, if not present Nielsen asks to set default
            @"channelName" : options[@"channelName"] ?: @"defaultChannelName",
            // if mediaURL is not available, Nielsen expects an empty value
            @"mediaURL" : options[@"mediaUrl"] ?: @""
        };

        [self startPlayheadTimer:payload];
        [self.nielsen play:channelInfo];
        SEGLog(@"[NielsenAppApi play: %@]", channelInfo);
        return;
    }

    if ([payload.event isEqualToString:@"Video Playback Paused"] ||
        [payload.event isEqualToString:@"Video Playback Seek Started"] ||
        [payload.event isEqualToString:@"Video Playback Buffer Started"] ||
        [payload.event isEqualToString:@"Video Playback Interrupted"] ||
        [payload.event isEqualToString:@"Video Playback Exited"]) {
        [self stopPlayheadTimer:payload];
        [self.nielsen stop];
        SEGLog(@"[NielsenAppApi stop]");
        return;
    }

    if ([payload.event isEqualToString:@"Video Playback Completed"]) {
        [self stopPlayheadTimer:payload];
        [self.nielsen end];
        SEGLog(@"[NielsenAppApi end]");
        return;
    }

#pragma mark - Content Events

    if ([payload.event isEqualToString:@"Video Content Started"]) {
        NSDictionary *contentMetadata = returnMappedContentProperties(properties, options, self.settings);
        [self.nielsen loadMetadata:contentMetadata];
        SEGLog(@"[NielsenAppApi loadMetadata:%@]", contentMetadata);
        [self startPlayheadTimer:payload];
        return;
    }

    if ([payload.event isEqualToString:@"Video Content Playing"]) {
        [self startPlayheadTimer:payload];
        return;
    }

    if ([payload.event isEqualToString:@"Video Content Completed"]) {
        [self stopPlayheadTimer:payload];
        [self.nielsen stop];
        return;
    }

#pragma mark - Ad Events

    if ([payload.event isEqualToString:@"Video Ad Started"]) {
        NSDictionary *adMetadata = returnMappedAdProperties(properties, options, self.settings);

        // In case of ad `type` preroll, call `loadMetadata` with metadata values for content, followed by `loadMetadata` with ad (preroll) metadata
        if ([properties[@"type"] isEqualToString:@"pre-roll"]) {
            NSDictionary *contentProperties = [properties valueForKey:@"content"];
            NSDictionary *adContentMetadata = returnMappedContentProperties(contentProperties, options, self.settings);
            [self.nielsen loadMetadata:adContentMetadata];
            SEGLog(@"[NielsenAppApi loadMetadata:%@]", adContentMetadata);
        }

        [self.nielsen loadMetadata:adMetadata];
        SEGLog(@"[NielsenAppApi loadMetadata:%@]", adMetadata);
        [self startPlayheadTimer:payload];
        return;
    }

    if ([payload.event isEqualToString:@"Video Ad Playing"]) {
        [self startPlayheadTimer:payload];
        return;
    }

    if ([payload.event isEqualToString:@"Video Ad Completed"]) {
        [self stopPlayheadTimer:payload];
        [self.nielsen stop];
        return;
    }
}

/**
 @return Opt-out URL string from the Nielsen App API to display in a web view.
 */
-(NSString *)optOutURL
{
    return [self.nielsen optOutURL];
}

/**
 @param urlString URL string from user's action to denote opt-out status for the Nielsen App API. Should be one of `nielsenappsdk://1` or `nielsenappsdk://0` for opt-out and opt-in, respectively
 @seealso https://engineeringportal.nielsen.com/docs/DTVR_iOS_SDK#The_legacy_opt-out_method_works_as_follows:
 */
-(void)userOptOutStatus:(NSString *)urlString
{
    [self.nielsen userOptOut:urlString];
}


#pragma mark - Screen

- (void)screen:(SEGScreenPayload *)payload
{
    NSDictionary *options = [payload.integrations valueForKey:@"nielsen-dcr"];
    NSDictionary *properties = payload.properties;
    NSDictionary *settings = self.settings;

    NSDictionary *metadata = @{
        @"type" : @"static",
        @"assetid" : returnCustomContentAssetId(properties, @"asset_id", settings),
        @"section" : returnCustomSectionProperty(properties, payload.name, settings),
        @"segA" : options[@"segA"] ?: @"",
        @"segB" : options[@"segB"] ?: @"",
        @"segC" : options[@"segC"] ?: @"",
        @"crossId1" : options[@"crossId1"] ?: @""
    };
    [self.nielsen loadMetadata:metadata];
    SEGLog(@"[NielsenAppApi loadMetadata:%@]", metadata);
}

@end
